### ISSUE TITLE
[MCOMPILER-417] Document compileSourceRoots, multiReleaseOutput

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-compiler-plugin#638]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** opened **[MCOMPILER-417](https://issues.apache.org/jira/browse/MCOMPILER-417?redirect=false)** and commented

Please update https://maven.apache.org/plugins/maven-compiler-plugin/compile-mojo.html and https://maven.apache.org/plugins/maven-compiler-plugin/multirelease.html  to mention the "compileSourceRoots", "multiReleaseOutput" parameters as well as providing an example like https://github.com/apache/maven-compiler-plugin/blob/master/src/it/multirelease-patterns/singleproject-runtime/pom.xml#L102-L108

in the usage section. It is an excellent solution that (seems to) work a lot cleaner than the other approaches.

 

---

**Affects:** 3.8.1

**Remote Links:**
- [GitHub Pull Request #43
](https://github.com/apache/maven-compiler-plugin/pull/43)

2 votes, 2 watchers

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Most documentation is already updated. I'm not going to expose `compileSourceRoots` as it an exclusive workaround for preparing a mrjar.

If there are extra source directories, one should use the [build-helper:add-source](https://www.mojohaus.org/build-helper-maven-plugin/add-source-mojo.html) to ensure that other plugins depending on source folders will use the same set.

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

Why doesn't https://maven.apache.org/plugins/maven-compiler-plugin/multirelease.html mention "multiReleaseOutput"?

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Where would you expect it? AFAIK the examples use it.

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

Where? I checked all the links coming off that page and the only mention of it I saw was in the integration tests (last link of the entire page). I don't think many users will find it buried in there.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Yes, it is a long page, but also an underestimated topic. Based on that you want something on top of the page like this:

```
TLDR;

Add an execution-block for every Java version looking like this:
                  <configuration>
                    <release>11</release>
                    <compileSourceRoots>
                      <compileSourceRoot>${project.basedir}/src/main/java11</compileSourceRoot>
                    </compileSourceRoots>
                    <multiReleaseOutput>true</multiReleaseOutput>
                  </configuration>
```

However, this isn't always true! It depends on the pattern you chose to test.

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

I understand, but I looked at the example code provided for each pattern and I didn't see any of them using the aforementioned code snippet. Did I miss anything? Or is this pattern missing from the list?

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Right, only singleproject-runtime uses it. Most others make normal jars, which are assemblied afterwards, moving the classes to the right directory. Again, it all depends on the pattern.

**Comment by jira-importer** [comment]
**[Gili](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cowwoc)** commented

The "Single Project" heading links to https://in.relation.to/2017/02/13/building-multi-release-jars-with-maven/ 

Is there any way to get that document to use `multiReleaseOutput` in place of `maven-antrun-plugin` and `maven-resources-plugin` as it does now?
